File: app/src/main/java/com/qianlei/jiaowu/ui/fragment/score/ScoreFragment.java
package com.qianlei.jiaowu.ui.fragment.score;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.qianlei.jiaowu.MainApplication;
import com.qianlei.jiaowu.R;
import com.qianlei.jiaowu.common.Result;
import com.qianlei.jiaowu.common.ResultType;
import com.qianlei.jiaowu.databinding.FragmentScoreBinding;
import com.qianlei.jiaowu.entity.Score;
import com.qianlei.jiaowu.utils.DateUtil;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import static androidx.recyclerview.widget.DividerItemDecoration.VERTICAL;

/**
 * 显示成绩的fragment
 *
 * @author qianlei
 */
public class ScoreFragment extends Fragment implements AdapterView.OnItemSelectedListener, SwipeRefreshLayout.OnRefreshListener {

    private ScoreViewModel scoreViewModel;
    private FragmentScoreBinding binding;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_score, container, false);

        binding.swipeRefreshLayout.setOnRefreshListener(this);
        binding.scoreTermChooseView.setItemSelectedListener(this);
        binding.recycleScore.setLayoutManager(new LinearLayoutManager(getContext()));
        binding.recycleScore.addItemDecoration(new DividerItemDecoration(MainApplication.getInstance(), VERTICAL));
        ViewModelProvider.AndroidViewModelFactory factory = new ViewModelProvider.AndroidViewModelFactory(MainApplication.getInstance());
        scoreViewModel = factory.create(ScoreViewModel.class);
        scoreViewModel.getResult().observe(this.getViewLifecycleOwner(), this::updateScore);
        binding.setLifecycleOwner(this);
        return binding.getRoot();
    }

    /**
     * 更新成绩数据
     *
     * @param result 更新后的成绩数据
     */
    private void updateScore(@NotNull Result<List<Score>> result) {
        if (result.getType() == ResultType.OK) {
            ScoreAdapter adapter = new ScoreAdapter(result.getData());
            binding.recycleScore.setAdapter(adapter);
        } else {
            Toast.makeText(getContext(), result.getMsg(), Toast.LENGTH_SHORT).show();
        }
        binding.swipeRefreshLayout.setRefreshing(false);
    }

    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        scoreViewModel.changeTerm(binding.scoreTermChooseView.getYear(), binding.scoreTermChooseView.getTerm());
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {
        scoreViewModel.changeTerm(String.valueOf(DateUtil.getCurYear()), String.valueOf(DateUtil.getCurTerm()));
    }

    @Override
    public void onRefresh() {
        scoreViewModel.refreshData(binding.scoreTermChooseView.getYear(), binding.scoreTermChooseView.getTerm());
    }
}

File: app/src/main/java/com/qianlei/jiaowu/ui/fragment/score/ScoreFragment.kt
package com.qianlei.jiaowu.ui.fragment.score

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener
import com.qianlei.jiaowu.R
import com.qianlei.jiaowu.common.Result
import com.qianlei.jiaowu.entity.Score
import com.qianlei.jiaowu.utils.TermUtil
import kotlinx.android.synthetic.main.fragment_exam.swipeRefreshLayout
import kotlinx.android.synthetic.main.fragment_score.*

/**
 * 显示成绩的fragment
 *
 * @author qianlei
 */
class ScoreFragment : Fragment(), OnItemSelectedListener, OnRefreshListener {
    private  val scoreViewModel by viewModels<ScoreViewModel>()

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val root = inflater.inflate(R.layout.fragment_score, container, false)
        scoreViewModel.scoreData.observe(this.viewLifecycleOwner, Observer { result: Result<List<Score>> -> updateScore(result) })
        return root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        recycleView.layoutManager = LinearLayoutManager(context)
        recycleView.addItemDecoration(DividerItemDecoration(context, DividerItemDecoration.VERTICAL))
        swipeRefreshLayout.setOnRefreshListener(this)
        swipeRefreshLayout.setColorSchemeResources(R.color.colorPrimary)
        scoreTermChooseView.setItemSelectedListener(this)
    }



    /**
     * 更新成绩数据
     *
     * @param result 更新后的成绩数据
     */
    private fun updateScore(result: Result<List<Score>>) {
        if (result.isSuccess()) {
            val data = result.data
            val adapter = ScoreAdapter(data)
            recycleView.adapter = adapter
        } else {
            Toast.makeText(context, result.msg, Toast.LENGTH_SHORT).show()
        }
        swipeRefreshLayout.isRefreshing = false
    }

    override fun onItemSelected(parent: AdapterView<*>?, view: View, position: Int, id: Long) {
        swipeRefreshLayout.isRefreshing = true
        scoreViewModel.changeTerm(scoreTermChooseView.term)
    }

    override fun onNothingSelected(parent: AdapterView<*>?) {
        swipeRefreshLayout.isRefreshing = true
        scoreViewModel.changeTerm(TermUtil.getNowTerm())
    }

    override fun onRefresh() {
        scoreViewModel.refreshData(scoreTermChooseView.term)
    }
}